tests/tests_firewall_fact.yml

```diff
@@ -0,0 +1,128 @@
+          vars:
+            custom: custom
+          fail:
+            msg: custom should not exist if there is no custom configuration
+          when: custom in firewall_config.keys()
+
+        - name: Fail if firewall_config defaults do not have attributes
+          fail:
+            msg: default {{ item }} does not have attributes
+          when: item | length == 0
+          loop: "{{ firewall_config.default | dict2items }}"
+
+        - name: save default ansible fact value
+          set_fact:
+            __previous_firewall_config: "{{ firewall_config }}"
+          no_log: true
+
+        # Test fact after adding custom settings
+
+        - name: modify firewalld configuration
+          shell:
+            cmd: |
+              firewall-cmd --permanent --add-service https
+              firewall-cmd --permanent --new-service custom
+
+        - name: refetch firewall_config
+          include_role:
+            name: linux-system-roles.firewall
+          vars:
+            firewall:
+
+        - name: Fail if firewall_config not changed
+          fail:
+            msg: >-
+              firewall_config was not modified
+              after changing firewall settings
+          when: firewall_config == __previous_firewall_config
+
+        - name: Fail if config defaults changed
+          fail:
+            msg: firewall_config defaults changed (should not have)
+          when: firewall_config.default != __previous_firewall_config.default
+
+        - name: Fail if custom config unchanged
+          vars:
+            custom: custom
+            zones: zones
+            services: services
+          fail:
+            msg: firewall customs did not change (when they should have)
+          when: >-
+            custom not in firewall_config.keys() or
+            zones not in firewall_config.custom.keys() or
+            services not in firewall_config.custom.keys()
+
+        - name: store previous firewall_config
+          set_fact:
+            __previous_firewall_config: "{{ firewall_config }}"
+          no_log: true
+
+        # Test detailed mode
+
+        - name: get firewall_config with detailed on
+          include_role:
+            name: linux-system-roles.firewall
+          vars:
+            firewall:
+              detailed: yes
+
+        - name: Create empty detailed defaults iterable
+          set_fact:
+            __default_config_iterable: []
+
+        - name: Populate __default_config_iterable
+          set_fact:
+            __default_config_iterable: "{{ __default_config_iterable +  item }}"
+          loop: "{{ firewall_config.default.values() | list }}"
+
+        - name: Fail if default is not list of dictionaries
+          fail:
+            msg: >-
+              detailed firewall_config.default not formatted properly -
+              see README.md for expected formatting
+          when: item is not mapping
+          loop: "{{ __default_config_iterable }}"
```

Comment:
can't you use `loop: "{{ firewall_config.default.values() }}"` and avoid the `set_fact` tasks?

Reply:
hmm - so the issue is that when you use `loop: "{{ firewall_config.default.values() }}"`, each `item` looks like this:
```
- key: value
  key1: value1
...
```
and you want `item` to look like this:
```
key: value
key1: value1
```
?

If so, can you use the `flatten` filter?  https://docs.ansible.com/ansible/2.9/user_guide/playbooks_filters.html#list-filters
`loop: "{{ firewall_config.default.values() | flatten }}"`

Reply:
Oh, that would work, thank you.